provider/provider.go

```diff
@@ -785,42 +786,64 @@ func (p *Provider) DispatchWithCtx(ctx context.Context, appPublicKey, chain stri
 }
 
 // Relay does request to be relayed to a target blockchain
-func (p *Provider) Relay(rpcURL string, input *RelayInput, options *RelayRequestOptions) (*RelayOutput, error) {
+func (p *Provider) Relay(rpcURL string, input *RelayInput, options *RelayRequestOptions) (*RelayOutput, *RelayOutputErr) {
 	return p.RelayWithCtx(context.Background(), rpcURL, input, options)
 }
 
 // RelayWithCtx does request to be relayed to a target blockchain
-func (p *Provider) RelayWithCtx(ctx context.Context, rpcURL string, input *RelayInput, options *RelayRequestOptions) (*RelayOutput, error) {
+func (p *Provider) RelayWithCtx(ctx context.Context, rpcURL string, input *RelayInput, options *RelayRequestOptions) (*RelayOutput, *RelayOutputErr) {
 	rawOutput, reqErr := p.doPostRequest(ctx, rpcURL, input, ClientRelayRoute, http.Header{})
 
 	defer closeOrLog(rawOutput)
 
 	if reqErr != nil && !errors.Is(reqErr, errOnRelayRequest) {
-		return nil, reqErr
+		return nil, &RelayOutputErr{Error: reqErr, StatusCode: rawOutput.Status}
 	}
 
-	bodyBytes, err := ioutil.ReadAll(rawOutput.Body)
+	bodyBytes, err := io.ReadAll(rawOutput.Body)
 	if err != nil {
-		return nil, err
+		return nil, &RelayOutputErr{Error: err, StatusCode: rawOutput.Status}
 	}
 
 	if errors.Is(reqErr, errOnRelayRequest) {
-		return nil, parseRelayErrorOutput(bodyBytes, input.Proof.ServicerPubKey)
+		return nil, &RelayOutputErr{Error: parseRelayErrorOutput(bodyBytes, input.Proof.ServicerPubKey), StatusCode: rawOutput.Status}
 	}
 
 	return parseRelaySuccesfulOutput(bodyBytes)
 }
 
-func parseRelaySuccesfulOutput(bodyBytes []byte) (*RelayOutput, error) {
+// TODO: REMOVE THIS HACKY FUNCTION
+func extractStatusFromResponse(response string) string {
+	patterns := []string{
```

Comment:
I'd precompile these.

```go
// Pre-compile the regular expressions for efficiency
var regexPatterns = []*regexp.Regexp{
	regexp.MustCompile(`"code"\s*:\s*(\d+)`),       // Matches `"code": 4XX` or similar if came from the node
	regexp.MustCompile(`(\d+)\s+Not Found`),        // Matches `404 Not Found` or similar
	regexp.MustCompile(`(\d+)\s+page not found`),   // Additional pattern for matching `404 page not found`
	regexp.MustCompile(`HTTP\/\d\.\d\s+(\d+)`),     // Matches `HTTP/1.1 4XX` or similar, for HTTP responses
	regexp.MustCompile(`"statusCode"\s*:\s*(\d+)`), // Matches `"statusCode": 4XX` or similar
}
```

Reply:
You can move it to an `init()` function in this file.

Reply:
I don't think we want to do that for now. This is a temporary function that will only be called when used the relay() function, and we have more than one function and many other cases were I don't see why we should load the regex

Reply:
But Relay is our must used function here, I think it could be worth

Reply:
There's no downside as far I can see.

Either you (regardless of whether its used in other place):

1. Recompile regexes on every call
2. Do it once